src/AggregateFunctions/AggregateFunctionSequenceNextNode.h

```diff
@@ -0,0 +1,473 @@
+
+#include <type_traits>
+
+
+namespace DB
+{
+
+const UInt32 MAX_EVENTS_SIZE = 64;
+
+/// NodeBase used to implement a linked list for storage of SequenceNextNodeImpl
+template <typename Node>
+struct NodeBase
+{
+    UInt64 size; /// size of payload
+
+    DataTypeDateTime::FieldType event_time;
+    std::bitset<MAX_EVENTS_SIZE> events_bitset;
+
+    char * data() { return reinterpret_cast<char *>(this) + sizeof(Node); }
+
+    const char * data() const { return reinterpret_cast<const char *>(this) + sizeof(Node); }
+
+    Node * clone(Arena * arena) const
+    {
+        return reinterpret_cast<Node *>(
+            const_cast<char *>(arena->alignedInsert(reinterpret_cast<const char *>(this), sizeof(Node) + size, alignof(Node))));
+    }
+
+    void write(WriteBuffer & buf) const
+    {
+        writeVarUInt(size, buf);
+        buf.write(data(), size);
+
+        writeBinary(event_time, buf);
+        UInt64 ulong_bitset = events_bitset.to_ulong();
+        writeBinary(ulong_bitset, buf);
+    }
+
+    static Node * read(ReadBuffer & buf, Arena * arena)
+    {
+        UInt64 size;
+        readVarUInt(size, buf);
+
+        Node * node = reinterpret_cast<Node *>(arena->alignedAlloc(sizeof(Node) + size, alignof(Node)));
+        node->size = size;
+        buf.read(node->data(), size);
+
+        readBinary(node->event_time, buf);
+        UInt64 ulong_bitset;
+        readBinary(ulong_bitset, buf);
+        node->events_bitset = ulong_bitset;
+
+        return node;
+    }
+};
+
+/// It stores String, timestamp, bitset of matched events.
+struct NodeString : public NodeBase<NodeString>
+{
+    using Node = NodeString;
+
+    static Node * allocate(const IColumn & column, size_t row_num, Arena * arena)
+    {
+        StringRef string = assert_cast<const ColumnString &>(column).getDataAt(row_num);
+
+        Node * node = reinterpret_cast<Node *>(arena->alignedAlloc(sizeof(Node) + string.size, alignof(Node)));
+        node->size = string.size;
+        memcpy(node->data(), string.data, string.size);
+
+        return node;
+    }
+
+    void insertInto(IColumn & column)
+    {
+        assert_cast<ColumnString &>(column).insertData(data(), size);
+    }
+
+    bool compare(const Node * rhs) const
+    {
+        auto cmp = strncmp(data(), rhs->data(), std::min(size, rhs->size));
+        return (cmp == 0) ? size < rhs->size : cmp < 0;
+    }
+};
+
+/// TODO : Expends SequenceNextNodeGeneralData to support other types
+template <typename Node, bool Descending>
+struct SequenceNextNodeGeneralData
+{
+    using Allocator = MixedAlignedArenaAllocator<alignof(Node *), 4096>;
+    using Array = PODArray<Node *, 32, Allocator>;
+
+    Array value;
+    bool sorted = false;
+
+    struct Comparator final
+    {
+        bool operator()(const Node * lhs, const Node * rhs) const
+        {
+            if constexpr (Descending)
+                return lhs->event_time == rhs->event_time ? !lhs->compare(rhs) : lhs->event_time > rhs->event_time;
```

Comment:
Query 
```
INSERT INTO test_sequenceNextNode_Nullable values ('1970-01-01 09:00:01',1,'A');
INSERT INTO test_sequenceNextNode_Nullable values ('1970-01-01 09:00:01',1,'A');

SELECT sequenceNextNode(1)(dt, action, action = 'A') AS next_node  FROM test_sequenceNextNode_Nullable;
```

leads to assertion error in debug mode `/contrib/libcxx/include/algorithm:814: _LIBCPP_ASSERT '!__comp_(__l, __r)' failed. Comparator does not induce a strict weak ordering`
We violate rule that `a < b` and `b < a` cannot return both `true`. https://stackoverflow.com/a/981299

I think we can leave condition `lhs->compare(rhs)` without negation, because `Decreasing` should affect only on time ordering, but if time is the same we can choose lowest string for decreasing and increasing.

How do you think?